### ISSUE TITLE
Possible ambiguous dispatch for identity_element on RealCircleGroup

### ISSUE THREAD

**Issue by dehann** [JuliaManifolds/Manifolds.jl#512]
Hi, the error stack seems suggests an ambiguous dispatch, I'm going to investigate further and find a way to work around this in the mean time.  Will post here again if I learn anything new:

```julia
  MethodError: identity_element(::AbstractManifold, ::Vector{Float64}) is ambiguous. Candidates:
    identity_element(::RealCircleGroup, p::AbstractArray) in Manifolds at /home/dehann/.julia/packages/Manifolds/0EI8E/src/groups/circle_group.jl:170
    identity_element(G::AbstractDecoratorManifold, p) in Manifolds at /home/dehann/.julia/packages/ManifoldsBase/gTRkF/src/nested_trait.jl:274
  To resolve the ambiguity, try making one of the methods more specific, or adding a new method more specific than any of the existing applicable methods.
  Stacktrace:
    [1] identity_element
      @ ~/.julia/packages/ManifoldsBase/gTRkF/src/nested_trait.jl:308 [inlined]
    [2] identity_element (repeats 4 times)
      @ ~/.julia/packages/ManifoldsBase/gTRkF/src/nested_trait.jl:282 [inlined]
    [3] identity_element
      @ ~/.julia/packages/ManifoldsBase/gTRkF/src/nested_trait.jl:275 [inlined]
    [4] #manikde!#51
```

Manifolds v0.8.21
ManifoldsBase v0.13.15

**Comment by kellertuer** [comment]
There is two things to note here.

1. The https://juliamanifolds.github.io/Manifolds.jl/latest/manifolds/group.html#Manifolds.CircleGroup iis the group on the complex circle, so its elements are just complex numbers and not vectors, for a vector of complex numbers where each entry is on the circle group, you need the power group(together with the power manifold of the circle)

2. yes that ambiguity is a bug (and the fix also makes your error vanish, but then you are still using the manifold not as intended).

**Comment by mateuszbaran** [comment]
There are fairly good reasons to represent points and tangent vectors on `RealCircleGroup` using vectors, primarily for compatibility with other libraries. IIRC I had some issues getting our circle code to work with ForwardDiff or Zygote with plain numbers or zero-index arrays.

**Comment by kellertuer** [comment]
Besides a little break in consistency ( currently – I hope – no other manifold implicitly also represents its own power manifold) – you then have to do quite some map implementations (which we removed maybe a year ago in favour of the PowerManifold / PowerGroup), my main argument is that we then have to adress all the ambiguities showing up, since we suddenly have to specialise on the points in our Layer 1.

**Comment by mateuszbaran** [comment]
I think the point isn't to make `RealCircleGroup` implicitly power but to allow length 1 `Vector`s as a representation. IIRC it generally works fairly well? And you've fixed the one issue with that in your PR?

**Comment by kellertuer** [comment]
With 1 vectors you mean vectors with one element, so `[0]`? That would be fine with me, sure.

**Comment by mateuszbaran** [comment]
Yes, I mean things like `[0]`.

**Comment by Affie** [comment]
> There are fairly good reasons to represent points and tangent vectors on RealCircleGroup using vectors, primarily for compatibility with other libraries. IIRC I had some issues getting our circle code to work with ForwardDiff or Zygote with plain numbers or zero-index arrays.

Yes, that is exactly why we use a length 1 vector `[0]`. We can't currently work with only the number as well as also struggling with `ForwardDiff` through Optim.

I suspect we will have the same issue with the `Quaternion` in `Unitary(1,ℍ)`

**Comment by mateuszbaran** [comment]
> Yes, that is exactly why we use a length 1 vector `[0]`. We can't currently work with only the number as well as also struggling with `ForwardDiff` through Optim.

I'm almost done with the ManifoldDiff.jl work and there is quite a bit of Manifolds.jl <-> ForwardDiff.jl compatibility, mostly seeding and differentiating through distance, log, exp and shortest_geodesic. If it would help then I think I could release the first bits this week.

**Comment by kellertuer** [comment]
I think that is a good idea when we have a first “good feature block” – maybe we can also find a few helping hand by announcing such an idea :)

**Comment by kellertuer** [comment]
> Edit: we use it in the same way as `TranslationGroup(1)`
> 
> ```julia
> julia> identity_element(TranslationGroup(1))
> 1-element Vector{Float64}:
>  0.0
> ```

I remember that we specifically have the zero-argument constructor for numbers and the `(1)` case for you situation :)

**Comment by mateuszbaran** [comment]
> I think that is a good idea when we have a first “good feature block” – maybe we can also find a few helping hand by announcing such an idea :)

ForwardDiff.jl and ChainRules.jl integration is basically ready at this point, I'm currently upstreaming changes to Manifolds.jl that are needed for our examples. Basically we just need to polish the examples at this point, and if the ForwardDiff.jl integration would bring some interest I think we can release it early :slightly_smiling_face: .

**Comment by kellertuer** [comment]
Do you want to do the examples as Pluto notebooks rendered in the docs as I introduced in Manopt.jl? I could help set that up in the docs.

**Comment by mateuszbaran** [comment]
I haven't used Pluto notebooks yet but if you think that's a good choice then we can surely use it here :slightly_smiling_face: .

**Comment by kellertuer** [comment]
The nice thing is that you can easily render the examples into the docs, the only disadvantage maybe that this only works after the package is registered.

**Comment by dehann** [comment]
> `[0]`

Acknowledged, thanks!

> If it would help then I think I could release the first bits this week.

Definitely going to take a look, and start figuring out how to use that for autodiff this side!

**Comment by dehann** [comment]
> yes that ambiguity is a bug

should I close this issue?

**Comment by kellertuer** [comment]
No, this is not needed, It will be closed (automatically) once we merge #513 :) because then this bug is resolved.
